### PR TITLE
clib, apply: Add apply feature to clib

### PR DESCRIPTION
**Pull request by qinqon**
Implment apply operation at clib also refactor the clib integration tests using acutest header as test framework

### CODE REVIEW

File: src/clib/lib.rs [hunk 1]
```diff
@@ -41,8 +42,48 @@ pub extern "C" fn nispor_net_state_retrieve(
...
+            },
+            Err(e) => unsafe {
+                *err_msg = CString::new(e.to_string()).unwrap().into_raw();
+                *err_kind = CString::new("bad format").unwrap().into_raw();
+                libc::EXIT_FAILURE
+            },
+        },
+        Err(e) => unsafe {
+            *err_msg = CString::new(e.to_string()).unwrap().into_raw();
+            *err_kind = CString::new("bad input").unwrap().into_raw();
```
Comment (cathay4t) [inline src/clib/lib.rs:86]: The `ErrorKind` has `Display` trait. So you should use `format!("{}", e.kind())` instead of hard-coded string `bad input`.

File: src/clib/nispor.h.in [hunk 2]
```diff
@@ -61,6 +61,33 @@ extern "C" {
...
+ * nispor_net_state_apply - Apply network state
+ *
+ * Version:
+ *      0.2
+ *
+ * Description:
+ *      Apply network state in the format of JSON.
+ *
+ * @state:
+ *      Input const char array for network state in json format.
```
Comment (cathay4t) [inline src/clib/nispor.h.in]: How about mention programe will quit by `assert` when input pointer is NULL.

File: test/clib/Makefile [hunk 3]
```diff
@@ -32,11 +32,21 @@ libnispor.so: $(DEBUG_NISPOR_SO_FILE)
...
+
+
 check: nispor_test
+	sudo \
+	RUST_BACKTRACE=1 \
 	LD_LIBRARY_PATH=$(TMPDIR) \
 		valgrind --trace-children=yes --leak-check=full \
 		--error-exitcode=1 \
-		./nispor_test 1>/dev/null
+		./nispor_test
```
Comment (cathay4t) [inline test/clib/Makefile:41]: This is unrelated to this PR.

File: test/clib/Makefile [hunk 4]
```diff
@@ -32,11 +32,21 @@ libnispor.so: $(DEBUG_NISPOR_SO_FILE)
 nispor_test: nispor_test.c nispor.h libnispor.so
 	$(CC) $(CFLAGS) $(LDFLAGS) -o nispor_test nispor_test.c $(LIBS)
 
+run: nispor_test
```
Comment (cathay4t) [inline test/clib/Makefile]: Why we need `make run` for this PR?

File: test/clib/acutest.h [hunk 5]
```diff
@@ -0,0 +1,1839 @@
...
+
+#ifdef _MSC_VER
+    #pragma warning(pop)
+#endif
+
+#ifdef __cplusplus
+    }  /* extern "C" */
+#endif
+
+#endif  /* #ifndef ACUTEST_H */
```
Comment (cathay4t) [inline test/clib/acutest.h]: I don't like bundling a header with 1.8K lines. Try something lighter and simpler. The clib of this project is quit small, not in the need of this complex giant.

**Review by cathay4t — CHANGES_REQUESTED** [review]
I don't like bundling acutest stuff. And clib of this project is not needing complex test setup.
Meanwhile, CI failed, please fix it also.

File: test/clib/Makefile (hunk 3 above)
Comment (qinqon) [reply test/clib/Makefile:41]: But without the output you don't see what is wrong not all the output is at stderr when there is an error.

File: test/clib/acutest.h (hunk 5 above)
Comment (qinqon) [reply test/clib/acutest.h]: This is test code, I don't see including test framework an issue, one thing we can do though is download the framework instead of storing it at the repo, what do you think ?

Comment (cathay4t) [reply test/clib/acutest.h]: I only see this bring risks like:
 * new dependency
 * maintenance effort required
 * hard to debug a giant C header
 
The C API is actually written in Rust, as long as we confirm we don't have memory leak and the function actually called the correct rust API,  that's enough which does not require complex test framework in C.

IMHO, this `acutest.h` brings zero benefit and a lot risks. So no, I don't approve this test framework in C.

Comment (qinqon) [reply test/clib/acutest.h]: ack, removing and simplifying.

File: src/clib/Cargo.toml [hunk 6]
```diff
@@ -15,3 +15,4 @@ crate-type = ["cdylib"]
 nispor = { path = "../lib" }
 libc = "0.2.74"
 serde_json = "1.0"
+serde_yaml = "0.8.17"
```
Comment (cathay4t) [inline src/clib/Cargo.toml]: Yaml is not considered as stable as JSON in python/go.

So as API, we should use json as we do for query.

**Review by cathay4t — CHANGES_REQUESTED** [review]

File: src/clib/Cargo.toml (hunk 6 above)
Comment (qinqon) [reply src/clib/Cargo.toml]: But the input is yaml we have to use the yaml parser, right ?

**Comment by qinqon** [comment]
@cathay4t CI is happy now, I was using an old rust version from fedora.

File: src/clib/Cargo.toml (hunk 6 above)
Comment (cathay4t) [reply src/clib/Cargo.toml]: The user of nispor API should convert yaml to json by their own tools. But as C API of nispor, we only support JSON.

There are some bad issue of YAML 1.2 recently fixed, that cause me lose the hope on supporting it as an API schema.
